### PR TITLE
Create the back-end for the new user registration #6 - BEHROUZ

### PR DESCRIPTION
**Pull request by bkarimii**
## Naming Rules

Name your PR like this: ISSUENUMBER-TITLE-YOURNAME

## Description

Write a summary of your change. Explain why you have made it.

## Related to

Make sure you include the issue number with a hash sign # so Github can link this PR to the right issue, like this:

Fixes #1

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have carefully reviewed my own code
- [x] I have commented my code
- [ ] I have updated any documentation

### CODE REVIEW

File: server/api.js [hunk 1]
```diff
@@ -63,6 +63,107 @@ router.get("/", (_, res) => {
...
+	);
+
+	for (let i = 4; i < length; i++) {
+		password += allChars.charAt(Math.floor(Math.random() * allChars.length));
+	}
+
+	password = password
+		.split("")
+		.sort(() => 0.5 - Math.random())
+		.join("");
```
Comment (soonick) [inline server/api.js]: What does this do?

File: server/api.js [hunk 2]
```diff
@@ -63,6 +63,107 @@ router.get("/", (_, res) => {
...
+	return password;
+}
+
+router.post("/signup", async (req, res) => {
+	const {
+		email,
+		userType,
+		firstName,
+		lastName,
+		userName,
```
Comment (soonick) [inline server/api.js]: We don't need usernames e-mails will be used

File: server/api.js [hunk 3]
```diff
@@ -63,6 +63,107 @@ router.get("/", (_, res) => {
...
+		company,
+		address,
+		description,
+	} = req.body;
+
+	if (!email || !userType || !firstName || !lastName) {
```
Comment (soonick) [inline server/api.js]: Do all validations before doing any queries and return all errors found in an array

File: server/api.js [hunk 4]
```diff
@@ -63,6 +63,107 @@ router.get("/", (_, res) => {
...
+			userTableValues = [userId, company, description, address];
+		} else {
+			await client.query("ROLLBACK");
+			return res.status(400).json({ message: "Unknown userType" });
+		}
+
+		await client.query(userTableQuery, userTableValues);
+
+		await client.query("COMMIT");
+		res.status(201).json({ message: "User registered successfully" });
```
Comment (soonick) [inline server/api.js]: The message is not necessary, the `201` is enough

File: server/api.js [hunk 5]
```diff
@@ -63,6 +63,107 @@ router.get("/", (_, res) => {
...
+	} catch (error) {
+		await client.query("ROLLBACK");
+		res.status(500).json({ message: "Server error: " + error.message });
```
Comment (soonick) [inline server/api.js]: Don't send server errors to clients, it could be a security issue.
```suggestion
		res.status(500).json({ code: "SERVER_ERROR" });
```

**Review by soonick — CHANGES_REQUESTED** [review]

File: server/api.js (hunk 1 above)
Comment (bkarimii) [reply server/api.js]: It made the passowrd more random after creating it. Should I keep it ?

Comment (soonick) [reply server/api.js]: What does it do? Randomizes the characters in the password? Seems unnecessary since the password is already randomly generated. It would be more useful to make it longer

File: server/api.js (hunk 2 above)
Comment (bkarimii) [reply server/api.js]: If I remove username from the code We also need to alter bidder table in the DB . Should I change the table and DB too ?

Comment (soonick) [reply server/api.js]: Let's do it in another PR

File: server/api.js [hunk 6]
```diff
@@ -57,12 +57,129 @@ const auth = async (req, res, next) => {
 	}
 };
 
-router.use(auth);
+// router.use(auth);
+const authRout = Router();
+authRout.use(auth);
```
Comment (soonick) [inline server/api.js]: Undo these changes

File: server/api.js [hunk 7]
```diff
@@ -57,12 +57,129 @@ const auth = async (req, res, next) => {
...
+
+	if (userType === "contractor" && (!company || !description || !address)) {
+		errors.push(
+			"Company, description, and address are required for contractors"
+		);
+	}
+
+	if (errors.length > 0) {
+		return res.status(400).json({
+			message: "Validation errors",
```
Comment (soonick) [inline server/api.js]:
```suggestion
			code: "VALIDATION_ERROR",
```

File: server/api.js [hunk 8]
```diff
@@ -57,12 +57,129 @@ const auth = async (req, res, next) => {
...
+			userTableQuery =
+				"INSERT INTO bidder (user_id, first_name, last_name, last_update) VALUES ($1, $2, $3, NOW())";
+			userTableValues = [userId, firstName, lastName];
+		} else if (userType === "contractor") {
+			userTableQuery =
+				"INSERT INTO contractor (user_id, company, description, address, last_update) VALUES ($1, $2, $3, $4, NOW())";
+			userTableValues = [userId, company, description, address];
+		} else {
+			await client.query("ROLLBACK");
+			return res.status(400).json({});
```
Comment (soonick) [inline server/api.js]: We already validated that the type has to be `bidder` or `contractor` if we get here, it's an error on our side (`500`)

File: server/api.js [hunk 9]
```diff
@@ -345,7 +462,7 @@ router.post("/bid/:bidId/status", async (req, res) => {
 		if (client) {
 			await client.query("ROLLBACK");
 		}
-		return res.status(500).send({ code: "SERVER_ERROR", error: error.message });
+		return res.status(500).send({ code: "SERVER_ERROR" });
```
Comment (soonick) [inline server/api.js:458]: Thanks for fixing this

**Review by soonick — CHANGES_REQUESTED** [review]

File: server/api.js [hunk 10]
```diff
@@ -63,6 +63,115 @@ router.get("/", (_, res) => {
...
+		errors.push("Last name is required");
+	}
+
+	if (userType === "buyer" && (!company || !description || !address)) {
+		errors.push("Company, description, and address are required for buyers");
+	}
+
+	if (errors.length > 0) {
+		return res.status(400).json({
+			code: "VALIDATION_ERROR",
```
Comment (soonick) [inline server/api.js:129]: You also need to send the errors

**Review by soonick — CHANGES_REQUESTED** [review]

File: server/api.js (hunk 10 above)
Comment (bkarimii) [reply server/api.js:129]: I thought it was not safe we send them back, but if we are going to do it, we can refer each message to a field or key then user knows what the error exactly is.
{field: "firstName", message: "first name is required" }.

Comment (soonick) [reply server/api.js:129]: It's safe in this case, because these are validation errors. Internal server errors should not be sent to clients

Comment (soonick) [reply server/api.js:129]: The way you had it before was fine:

```
errors: errors
```

**Review by soonick — APPROVED** [review]
